runtime/src/main/java/org/corfudb/runtime/LogReplicationRoutingQueueListener.java

```diff
@@ -0,0 +1,357 @@
+
+
+        if (tableNames.contains(REPLICATION_STATUS_TABLE_NAME)) {
+            Preconditions.checkState(results.getEntries().keySet().size() == 1,
+                    "Replication Status Table Update received with other tables");
+            processReplicationStatusUpdate(results);
+            return;
+        }
+
+        // Data Updates
+        if (clientFullSyncPending.get()) {
+            // If the listener started when snapshot sync was ongoing, ignore all data updates until it ends.  When
+            // it ends, the client will perform a full sync and build a consistent state containing these updates.
+            return;
+        }
+
+        boolean callbackResult = false;
+        List<Queue.RoutingTableEntryMsg> entries =
+                results.getEntries().entrySet().stream().map(CSEntry ->
+                        (Queue.RoutingTableEntryMsg)CSEntry.getValue().get(0).getPayload()).collect(Collectors.toList());
+        if (snapshotSyncInProgress.get()) {
+            // For routing queue listener, we have only 1 entry inside corfuStreamEntries.
+            callbackResult = processUpdatesInSnapshotSync(entries);
+        } else {
+            callbackResult = processUpdatesInLogEntrySync(entries);
+        }
+        if (callbackResult) {
+            // When the client processed the entry successfully, delete the entry from the routing queue.
+            deleteEntryFromQueue(results);
+        }
+        this.lastProcessedEntryTs = results.getTimestamp();
+    }
+
+    private void deleteEntryFromQueue(CorfuStreamEntries results) {
+        Map<TableSchema, List<CorfuStreamEntry>> entries = results.getEntries();
+        List<CorfuStreamEntry> routingQueueTableEntries =
+                entries.entrySet().stream().filter(
+                            e -> e.getKey().getTableName().equals(LogReplicationUtils.REPLICATED_QUEUE_NAME_PREFIX))
+                        .map(Map.Entry::getValue)
+                        .findFirst()
+                        .get();
+
+        for (CorfuStreamEntry entry : routingQueueTableEntries) {
+            Queue.CorfuGuidMsg key = (Queue.CorfuGuidMsg)entry.getKey();
+            // Only process updates where operation type == UPDATE.
+            if (entry.getOperation() == CorfuStreamEntry.OperationType.UPDATE) {
+                // TODO: Check payload type??
+                //  Queue.RoutingTableEntryMsg payload = (Queue.RoutingTableEntryMsg)entry.getPayload();
+                try (TxnContext tx = corfuStore.txn(namespace)) {
+                    Table<Queue.CorfuGuidMsg, Queue.RoutingTableEntryMsg, Queue.CorfuQueueMetadataMsg> routingQueue;
+                    // TODO: Get the table from CorfuStreamEntries schema or registry table?
+                    routingQueue =
+                            corfuStore.openQueue(namespace, LogReplicationUtils.REPLICATED_QUEUE_NAME_PREFIX,
```

Comment:
we can open the queue once in the constructor of this class.  It need not be opened every time.